Add support of mulitple languages

dedifferentiator opened the pull request:
Hey! I added support of Russian and Haskell as alternative languages.

There are not really many quotes, because I didn't manage to find a decent dataset with russian quotes and that's annoying :c , and I kinda don't want to write a parser for that purpose yet (it's a bit difficult to filter good quotes), so I manually added some of them, gonna spend 30 min per day to add new ones for a month or so, I think having around ~500 of them would be nice for the start. Although I may give up and write a parser in the end...

dedifferentiator commented:
oops...CI seems to be broken

rmehri01 commented on src/Quotes.hs:
```diff
@@ -14,12 +14,29 @@ import Data.Aeson
+
+instance ToJSON QuotesSet
+
+instance Show QuotesSet where
+  show English = "English"
+  show Russian = "Russian"
+  show Haskell = "Haskell"
```
Isn't this the same thing as deriving `Show`?

dedifferentiator replied:
Probably yes for sure! but I wasn't too confident about it, because `Show` likes to add additonal `" "` to objects, and I didn't want to test it then. I'll put the fix

rmehri01 commented on src/Quotes.hs:
```diff
@@ -39,11 +56,26 @@ instance ToJSON Quote where
+-- for the given `QuotesSet`
+getFirstMatch :: QuotesSet -> [(FilePath, Maybe [a])] -> Maybe [a]
+getFirstMatch qs (x : xs) =
+  if fst x == getPath qs
+    then snd x
+    else getFirstMatch qs xs
   where
-    qs = fromMaybe (error "could not decode JSON into quote") mqs
-    mqs = decodeStrict $(embedFile "resources/quotes.json")
+    getPath set = case set of
+      English -> "quotes_en.json"
+      Russian -> "quotes_ru.json"
+      Haskell -> "quotes_hask.json"
+getFirstMatch _ [] = Nothing
```
This seems like the same thing as `find` from `Data.List`, perhaps:

```haskell
getFirstMatch :: QuotesSet -> [(FilePath, Maybe [a])] -> Maybe [a]
getFirstMatch qs xs = find (\(path, _) -> path == getPath qs) xs >>= snd
  where
    getPath set = case set of
      English -> "quotes_en.json"
      Russian -> "quotes_ru.json"
      Haskell -> "quotes_hask.json"
```

Might be a bit nicer?

rmehri01 commented on src/UI/Online.hs:
```diff
@@ -116,31 +120,42 @@ handleKeyOnline s ev = case s of
-    StartGame q gs -> M.continue (OnlineGame (OnlineGameState (initializeGameState q) room (localSt ^. username) conn gs))
+    RoomUpdate rs -> M.continue . WaitingRoom $ WaitingRoomState room qs localSt conn rs
+    StartGame q gs -> M.continue . OnlineGame $ OnlineGameState (initializeGameState qs q) room qs (localSt ^. username) conn gs
+    -- FIXME :)
     _ -> undefined
```
This is supposed to be undefined since the other cases aren't supposed to happen. Do you mean there should be a better way to represent it?

dedifferentiator replied:
I agree, just don't feel like it's a good thing to leave undefined somewhere in code. Maybe replace it with `error "Error: unexpected game state"?

rmehri01 replied:
Ah okay fair enough, that makes sense!

rmehri01 reviewed (commented):
Hi! Wow this is a lot of work, so thanks again! I tried it locally and it seems to work good, have you tried the online locally? I haven't quite got around to it yet because I've been busy with school and other work but I'll try to get to it soon. 

As for the other quotes, do you know how well something like google translate would work because it would be pretty easy to write a script for it but I don't know how good the quotes will translate. 

Also what do you mean by writing a parser for it? Maybe if it is a better way to do it I could try to help out!

rmehri01 commented:
> oops...CI seems to be broken

Hmm it seems like the error is the same as [this one](https://github.com/domdere/haskell-lint-action/issues/9). Maybe I will have to find another way to do linting or remove it entirely.

dedifferentiator commented:
>  have you tried the online locally?

Yes, it seems to work, but you'd better check it out once more :)

> I don't know how good the quotes will translate.

Pretty awfully, unfortunately, it would be around 60% +- ok quality with one-sentence quotes, but with number of sentences the quality drops significantly due to slang, names, etc, I don't think that translating the quotes is an option.

> Also what do you mean by writing a parser for it?

I mean to write a web crawler, which would go through any [livelib](https://www.livelib.ru/book/1000005849/quotes-gedel-esher-bah-eta-beskonechnaya-girlyanda-daglas-hofshtadter) stuff or citaty.info and collect quotes from them. Generally the problem is how to sort the good quotes from the err... _cringy_ ones. Maybe the good strategy would be to gather links with good books in one place, run parser on them, replace unprintable on default layout symbols like (`—` to `-` ) and then create a script that would show a quote and you just have to press y/n, add a quote to json or skip it. Still not perfect, but already more feasible, than doing it manually.

What about haskell source code quotes, I'm not so sure, cause it's almost no difference what to type to me, and as I see it, it's to have 2-3 quotes from different projects repos, and the number of these quotes in the dataset doesn't matter much. What are your thoughts about that?

rmehri01 commented:
> Maybe the good strategy would be to gather links with good books in one place, run parser on them, replace unprintable on default layout symbols like (— to - ) and then create a script that would show a quote and you just have to press y/n, add a quote to json or skip it.

Oh okay, I see what you mean. That seems like a pretty good way to do it, especially for many different quotes!

> What about haskell source code quotes, I'm not so sure, cause it's almost no difference what to type to me, and as I see it, it's to have 2-3 quotes from different projects repos, and the number of these quotes in the dataset doesn't matter much. What are your thoughts about that?

Yeah, honestly it doesn't make much of a difference to me either so I think that it's good enough to just have a few like you said.

rmehri01 reviewed (commented):
Hi, just a few more things I ran into while testing the server!

When you try to enter a name or room with a 'q' in it, it just quits instead of entering it as text. I think that it would be easier to just reserve 'ESC' for quitting instead of having multiple bindings for it.

Also, when the client first joins the game, I don't think that the quote set is properly updated for them:

![image](https://user-images.githubusercontent.com/52933714/107157657-75338200-6942-11eb-855e-ea15e940e20d.png)

dedifferentiator commented:
> I think that it would be easier to just reserve 'ESC' for quitting instead of having multiple bindings for it.

Yes, I also noticed this bug and fixed it, you probably didn't pull last changes at https://github.com/rmehri01/thock/pull/13/commits/51c96e3bfa1184829f59e84ca39a83ff2f6df185

> Also, when the client first joins the game, I don't think that the quote set is properly updated for them:

Fixed :)

rmehri01 commented:
> Yes, I also noticed this bug and fixed it, you probably didn't pull last changes at 51c96e3

Oops, my bad!

rmehri01 reviewed (approved):
Looks good to me, thank you again and sorry about the wait!

dedifferentiator commented:
It's okay, no problem at all!
